samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt

```diff
@@ -224,43 +233,38 @@ abstract class DownloaderActivity : AppCompatActivity() {
                 emit(LoadStatus.FailedToLoad(it))
                 return@flow
             }
+            // set up the file at the download path
+            val destinationFilePath = provisionLocation.path + File.separator + portalItem.name
+            // set up the download URL
+            val downloadURL =
+                "${portalItem.portal.url}/sharing/rest/content/items/${portalItem.itemId}/data"
             // get the data of the PortalItem
-            val portalItemData = portalItem.fetchData()
-            val byteArray = portalItemData.getOrElse {
-                // close dialog and emit status
-                loadingDialog.dismiss()
-                emit(LoadStatus.FailedToLoad(it))
-                return@flow
-            }
-            // get the byteArray of the PortalItem
-            runCatching {
-                val byteArrayInputStream = ByteArrayInputStream(byteArray)
-                val data = ByteArray(1024)
-                var downloadCount: Int
-                downloadCount = byteArrayInputStream.read(data)
-                while (downloadCount != -1) {
-                    downloadCount = byteArrayInputStream.read(data)
+            ArcGISEnvironment.arcGISHttpClient.download(
+                url = downloadURL,
+                destinationFile = File(destinationFilePath)
+            ) { totalBytes, bytesRead ->
+                if (totalBytes != null) {
+                    // update the download percentage progress
+                    val percentage = ((100.0 * bytesRead) / totalBytes).roundToInt()
+                    Handler(Looper.getMainLooper()).post {
+                        progressIndicator.progress = percentage
+                        progressTV.text = "$percentage%"
+                    }
                 }
-                // set up the file at the download path
-                val destinationFilePath = provisionLocation.path + File.separator + portalItem.name
-                val provisionFile = File(destinationFilePath)
-                // create file at location to write the PortalItem ByteArray
-                provisionFile.createNewFile()
-                // create and write the file output stream
-                val writeOutputStream = FileOutputStream(provisionFile)
-                writeOutputStream.write(byteArray)
-
+            }.onSuccess {
                 // unzip the file if it is a .zip
                 if (portalItem.name.contains(".zip")) {
+                    // set up the file at the download path
+                    val provisionFile = File(destinationFilePath)
+                    // set up the zip input stream
                     val fileInputStream = FileInputStream(destinationFilePath)
                     val zipInputStream = ZipInputStream(BufferedInputStream(fileInputStream))
```

Comment:
I believe all this unzipping logic should be implemented in a "try with resources" style, i.e. the input streams should be closed in a `finally` block. In Kotlin you would typically do this with the extension function [Closable.use](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html).

Reply:
I've updated the input streams to:
```kotlin
FileInputStream(destinationFilePath).use { fileInputStream ->
    ZipInputStream(BufferedInputStream(fileInputStream)).use { zipInputStream ->
        ...
    }
}
```